File: src/quickpack/metadata.py
"""Generate metadata files for a charm from ``charmcraft.yaml``."""

import datetime
import pathlib
import platform
import shutil
from typing import Any

import yaml

import quickpack

# Maps ``platform.machine()`` values to Juju architecture labels.
_MACHINE_TO_ARCH: dict[str, str] = {
    "x86_64": "amd64",
    "arm64": "arm64",
    "aarch64": "arm64",
    "armv7l": "armhf",
    "ppc64le": "ppc64el",
    "s390x": "s390x",
    "riscv64": "riscv64",
}


def local_arch() -> str:
    """Return the Juju architecture label for the current machine."""
    machine = platform.machine()
    try:
        return _MACHINE_TO_ARCH[machine]
    except KeyError:
        raise RuntimeError(f"Unsupported architecture: {machine}") from None


def parse_charmcraft_yaml(charm_dir: pathlib.Path) -> dict[str, Any]:
    """Load and return the parsed ``charmcraft.yaml``.

    If the ``name`` field is missing, it is inferred from the directory
    name (matching charmcraft's behaviour).
    """
    path = charm_dir / "charmcraft.yaml"
    if not path.exists():
        raise FileNotFoundError(f"charmcraft.yaml not found in {charm_dir}")
    with path.open(encoding="utf-8") as fh:
        data = yaml.safe_load(fh)
    if not isinstance(data, dict):
        raise ValueError("charmcraft.yaml must be a YAML mapping")
    if "name" not in data:
        data["name"] = charm_dir.name
    return data


def resolve_base(project: dict[str, Any]) -> tuple[str, str]:
    """Determine the (distro, series) base from the project config.

    Returns e.g. ``("ubuntu", "24.04")``.
    """
    # Modern ``base:`` field (e.g. ``base: "ubuntu@24.04"``).
    if (base_str := project.get("base")) and "@" in base_str:
        distro, series = base_str.split("@", 1)
        return distro, series

    # ``platforms:`` keys (e.g. ``ubuntu@24.04:amd64``).
    if platforms := project.get("platforms"):
        for key in platforms:
            if "@" in str(key):
                label = str(key).split(":")[0]  # Strip arch suffix.
                distro, series = label.split("@", 1)
                return distro, series

    # Legacy ``bases:`` format.
    if bases := project.get("bases"):
        for base_entry in bases:
            for run_on in base_entry.get("run-on", []):
                name = run_on.get("name", "ubuntu")
                channel = str(run_on.get("channel", "24.04"))
                return name, channel

    return "ubuntu", "24.04"


def resolve_entrypoint(project: dict[str, Any]) -> str:
    """Determine the charm entrypoint from the parts config."""
    for part in (project.get("parts") or {}).values():
        if ep := part.get("charm-entrypoint"):
            return ep
    return "src/charm.py"


def _validate_entrypoint(entrypoint: str, charm_dir: pathlib.Path) -> None:
    """Reject entrypoints the dispatch script could not safely launch.

    The entrypoint is interpolated into a generated POSIX shell script
    and is also the path the charm's process actually executes — so it
    must be a relative file that lives inside the charm tree, with no
    parent traversal and no shell-hostile characters.  Catching these
    cases at pack time turns a delayed deploy-time hook failure into a
    crisp build error.
    """
    if not isinstance(entrypoint, str):
        raise ValueError(f"charm-entrypoint must be a string, got {type(entrypoint).__name__}")
    if not entrypoint:
        raise ValueError("charm-entrypoint must not be empty")

    # Reject characters that would break the dispatch script's exec
    # line or smuggle additional commands.  Newlines, NULs, and quotes
    # are all immediate hazards; backslashes get rejected because we
    # do not want to reason about shell quoting at runtime.
    forbidden = set("\n\r\0\"'`\\$")
    bad = sorted(forbidden.intersection(entrypoint))
    if bad:
        raise ValueError(f"charm-entrypoint contains forbidden characters: {bad}")

    candidate = pathlib.PurePosixPath(entrypoint)
    if candidate.is_absolute():
        raise ValueError(f"charm-entrypoint must be relative, got {entrypoint!r}")
    if any(part == ".." for part in candidate.parts):
        raise ValueError(f"charm-entrypoint must stay inside the charm tree, got {entrypoint!r}")

    full = (charm_dir / entrypoint).resolve()
    try:
        full.relative_to(charm_dir.resolve())
    except ValueError as exc:
        raise ValueError(
            f"charm-entrypoint resolves outside the charm directory: {entrypoint!r}"
        ) from exc

    if not full.is_file():
        raise FileNotFoundError(f"charm-entrypoint points to a missing file: {entrypoint!r}")


def validate_project(project: dict[str, Any], charm_dir: pathlib.Path) -> None:
    """Validate the charmcraft.yaml fields the pack path depends on.

    Run before any heavy work so malformed metadata fails fast with a
    targeted message instead of triggering a confusing failure deep in
    parts processing or zip assembly.
    """
    name = project.get("name")
    if not isinstance(name, str) or not name.strip():
        raise ValueError("charmcraft.yaml: 'name' must be a non-empty string")

    parts = project.get("parts")
    if parts is not None and not isinstance(parts, dict):
        raise ValueError("charmcraft.yaml: 'parts' must be a mapping when set")

    _validate_entrypoint(resolve_entrypoint(project), charm_dir)


# Maps the Ubuntu series to the system-supplied CPython on that LTS.
# Used to pick the build venv's Python version so the resulting
# ``venv/lib/pythonX.Y`` directory matches what the unit's ``python3``
# will read at runtime.  Out-of-range or unknown series fall back to
# whatever Python the build host's ``python3`` resolves to — which is
# the historical behaviour, so existing quickpack runs that happened
# to match the host Python don't change.
_UBUNTU_PYTHON: dict[str, str] = {
    "20.04": "3.8",
    "22.04": "3.10",
    "24.04": "3.12",
    "24.10": "3.12",
    "25.04": "3.13",
    "26.04": "3.14",
}


def resolve_target_python(project: dict[str, Any]) -> str | None:
    """Return the CPython version label the unit will run, or ``None``.

    Reads the (optional) ``build-base`` field first, since that's the
    series the charm code is actually compiled and packed against; the
    runtime ``base`` falls through when ``build-base`` is omitted.
    Returns a string like ``"3.12"`` suitable for ``uv venv --python``,
    or ``None`` when the series is unknown so callers can fall back to
    the host's default.
    """
    build_base = str(project.get("build-base") or "").strip()
    if "@" in build_base:
        _, _, series = build_base.partition("@")
        if series in _UBUNTU_PYTHON:
            return _UBUNTU_PYTHON[series]

    distro, series = resolve_base(project)
    if distro != "ubuntu":
        return None
    return _UBUNTU_PYTHON.get(series)


def generate_metadata(project: dict[str, Any]) -> dict[str, Any]:
    """Build the ``metadata.yaml`` content from a parsed ``charmcraft.yaml``.

    Performs the same field renaming and link flattening that charmcraft does.
    """
    metadata: dict[str, Any] = {}

    # Direct-copy fields.
    for key in (
        "name",
        "summary",
        "description",
        "assumes",
        "containers",
        "devices",
        "extra-bindings",
        "peers",
        "provides",
        "requires",
        "resources",
        "storage",
        "subordinate",
        "terms",
    ):
        if key in project:
            metadata[key] = project[key]

    # Rename ``title`` → ``display-name``.
    if "title" in project:
        metadata["display-name"] = project["title"]

    # Flatten ``links`` into top-level metadata fields.
    if links := project.get("links"):
        if "documentation" in links:
            metadata["docs"] = links["documentation"]
        if "contact" in links:
            contact = links["contact"]
            if isinstance(contact, str):
                contact = [contact]
            metadata["maintainers"] = contact
        if "issues" in links:
            metadata["issues"] = links["issues"]
        if "website" in links:
            metadata["website"] = links["website"]
        if "source" in links:
            metadata["source"] = links["source"]

    return metadata


def _detect_language(project: dict[str, Any]) -> str:
    """Detect the charm's primary language from parts config.

    Returns ``"python"`` when a ``uv`` or ``charm`` plugin is used,
    ``"unknown"`` otherwise.
    """
    for part in (project.get("parts") or {}).values():
        if isinstance(part, dict) and part.get("plugin") in ("uv", "charm"):
            return "python"
    return "unknown"


def generate_manifest(
    project: dict[str, Any],
    arch: str | None = None,
) -> dict[str, Any]:
    """Build the ``manifest.yaml`` content."""
    distro, series = resolve_base(project)
    if arch is None:
        arch = local_arch()

    language = _detect_language(project)

    return {
        "charmcraft-version": f"quickpack-{quickpack.__version__}",
        "charmcraft-started-at": datetime.datetime.now(datetime.UTC).isoformat(),
        "bases": [
            {
                "name": distro,
                "channel": series,
                "architectures": [arch],
            },
        ],
        "analysis": {
            "attributes": [
                {"name": "language", "result": language},
                {"name": "framework", "result": "unknown"},
            ],
        },
    }


def _resolve_platform_label(project: dict[str, Any], arch: str) -> str:
    """Determine the platform label for the charm filename.

    Charmcraft uses the platform key directly:
    - ``platforms: {amd64: null}`` → label is ``amd64``
    - ``platforms: {ubuntu@24.04:amd64: null}`` → label is ``ubuntu@24.04-amd64``
      (colon replaced with hyphen)

    When no platform key matches, falls back to ``{base}-{arch}``.
    """
    if platforms := project.get("platforms"):
        for key in platforms:
            key_str = str(key)
            # Platform key that is just the arch name.
            if key_str == arch:
                return arch
            # Platform key that includes the base and arch.
            if key_str.endswith(f":{arch}") or key_str.endswith(f"-{arch}"):
                return key_str.replace(":", "-")
        # Use the first platform key if none matched explicitly.
        first = str(next(iter(platforms)))
        return first.replace(":", "-")

    distro, series = resolve_base(project)
    return f"{distro}@{series}-{arch}"


def charm_filename(project: dict[str, Any], arch: str | None = None) -> str:
    """Return the standard charm filename, e.g. ``myapp_amd64.charm``."""
    name = project["name"]
    if arch is None:
        arch = local_arch()
    label = _resolve_platform_label(project, arch)
    return f"{name}_{label}.charm"


def write_optional_yaml(
    project: dict[str, Any],
    field: str,
    filename: str,
    charm_dir: pathlib.Path,
    prime_dir: pathlib.Path,
) -> None:
    """Write ``actions.yaml`` or ``config.yaml`` into *prime_dir*.

    Prefers copying the source file from *charm_dir* if it exists on disk,
    otherwise generates it from the project dict.
    """
    source = charm_dir / filename
    dest = prime_dir / filename
    if source.is_file():
        shutil.copy2(str(source), str(dest))
    elif field in project and project[field]:
        with dest.open("w", encoding="utf-8") as fh:
            yaml.safe_dump(project[field], fh, default_flow_style=False)

File: src/quickpack-rs/src/metadata.rs
//! Generate metadata files for a charm from `charmcraft.yaml`.

use serde_yaml::Value;
use std::collections::BTreeMap;
use std::path::Path;

/// Maps `uname -m` values to Juju architecture labels.
pub fn local_arch() -> Result<String, String> {
    let machine = std::env::consts::ARCH;
    match machine {
        "x86_64" => Ok("amd64".to_string()),
        "aarch64" => Ok("arm64".to_string()),
        "arm" => Ok("armhf".to_string()),
        "powerpc64" => Ok("ppc64el".to_string()),
        "s390x" => Ok("s390x".to_string()),
        "riscv64" => Ok("riscv64".to_string()),
        _ => Err(format!("Unsupported architecture: {machine}")),
    }
}

/// Load and return the parsed `charmcraft.yaml`.
pub fn parse_charmcraft_yaml(charm_dir: &Path) -> Result<BTreeMap<String, Value>, String> {
    let path = charm_dir.join("charmcraft.yaml");
    if !path.exists() {
        return Err(format!(
            "charmcraft.yaml not found in {}",
            charm_dir.display()
        ));
    }
    let content =
        std::fs::read_to_string(&path).map_err(|e| format!("Failed to read charmcraft.yaml: {e}"))?;
    let data: Value =
        serde_yaml::from_str(&content).map_err(|e| format!("Invalid YAML: {e}"))?;
    match data {
        Value::Mapping(m) => {
            let mut result = BTreeMap::new();
            for (k, v) in m {
                if let Value::String(key) = k {
                    result.insert(key, v);
                }
            }
            // Infer name from directory if missing (matches charmcraft behaviour).
            if !result.contains_key("name") {
                if let Some(dir_name) = charm_dir.file_name().and_then(|n| n.to_str()) {
                    result.insert("name".to_string(), Value::String(dir_name.to_string()));
                }
            }
            Ok(result)
        }
        _ => Err("charmcraft.yaml must be a YAML mapping".to_string()),
    }
}

/// Determine the (distro, series) base from the project config.
pub fn resolve_base(project: &BTreeMap<String, Value>) -> (String, String) {
    // Modern `base:` field (e.g. `base: "ubuntu@24.04"`).
    if let Some(Value::String(base_str)) = project.get("base") {
        if let Some((distro, series)) = base_str.split_once('@') {
            return (distro.to_string(), series.to_string());
        }
    }

    // `platforms:` keys (e.g. `ubuntu@24.04:amd64`).
    if let Some(Value::Mapping(platforms)) = project.get("platforms") {
        for key in platforms.keys() {
            if let Value::String(key_str) = key {
                if key_str.contains('@') {
                    let label = key_str.split(':').next().unwrap_or(key_str);
                    if let Some((distro, series)) = label.split_once('@') {
                        return (distro.to_string(), series.to_string());
                    }
                }
            }
        }
    }

    // Legacy `bases:` format.
    if let Some(Value::Sequence(bases)) = project.get("bases") {
        for base_entry in bases {
            if let Value::Mapping(entry) = base_entry {
                if let Some(Value::Sequence(run_ons)) = entry.get(Value::String("run-on".into())) {
                    for run_on in run_ons {
                        if let Value::Mapping(ro) = run_on {
                            let name = ro
                                .get(Value::String("name".into()))
                                .and_then(|v| v.as_str())
                                .unwrap_or("ubuntu");
                            let channel = ro
                                .get(Value::String("channel".into()))
                                .map(|v| match v {
                                    Value::String(s) => s.clone(),
                                    _ => format!("{v:?}"),
                                })
                                .unwrap_or_else(|| "24.04".to_string());
                            return (name.to_string(), channel);
                        }
                    }
                }
            }
        }
    }

    ("ubuntu".to_string(), "24.04".to_string())
}

/// Determine the charm entrypoint from the parts config.
pub fn resolve_entrypoint(project: &BTreeMap<String, Value>) -> String {
    if let Some(Value::Mapping(parts)) = project.get("parts") {
        for part in parts.values() {
            if let Value::Mapping(part_map) = part {
                if let Some(Value::String(ep)) =
                    part_map.get(Value::String("charm-entrypoint".into()))
                {
                    return ep.clone();
                }
            }
        }
    }
    "src/charm.py".to_string()
}

/// Build the `metadata.yaml` content from a parsed `charmcraft.yaml`.
pub fn generate_metadata(project: &BTreeMap<String, Value>) -> BTreeMap<String, Value> {
    let mut metadata = BTreeMap::new();

    // Direct-copy fields.
    let direct_copy = [
        "name",
        "summary",
        "description",
        "assumes",
        "containers",
        "devices",
        "extra-bindings",
        "peers",
        "provides",
        "requires",
        "resources",
        "storage",
        "subordinate",
        "terms",
    ];
    for key in direct_copy {
        if let Some(val) = project.get(key) {
            metadata.insert(key.to_string(), val.clone());
        }
    }

    // Rename `title` -> `display-name`.
    if let Some(val) = project.get("title") {
        metadata.insert("display-name".to_string(), val.clone());
    }

    // Flatten `links` into top-level metadata fields.
    if let Some(Value::Mapping(links)) = project.get("links") {
        if let Some(v) = links.get(Value::String("documentation".into())) {
            metadata.insert("docs".to_string(), v.clone());
        }
        if let Some(v) = links.get(Value::String("contact".into())) {
            let contact = match v {
                Value::String(s) => Value::Sequence(vec![Value::String(s.clone())]),
                _ => v.clone(),
            };
            metadata.insert("maintainers".to_string(), contact);
        }
        if let Some(v) = links.get(Value::String("issues".into())) {
            metadata.insert("issues".to_string(), v.clone());
        }
        if let Some(v) = links.get(Value::String("website".into())) {
            metadata.insert("website".to_string(), v.clone());
        }
        if let Some(v) = links.get(Value::String("source".into())) {
            metadata.insert("source".to_string(), v.clone());
        }
    }

    metadata
}

/// Build the `manifest.yaml` content.
pub fn generate_manifest(
    project: &BTreeMap<String, Value>,
    arch: &str,
) -> BTreeMap<String, Value> {
    let (distro, series) = resolve_base(project);
    let now = chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Micros, true);

    let mut manifest = BTreeMap::new();
    manifest.insert(
        "charmcraft-version".to_string(),
        Value::String(format!("quickpack-{}", env!("CARGO_PKG_VERSION"))),
    );
    manifest.insert("charmcraft-started-at".to_string(), Value::String(now));

    let mut base_entry = serde_yaml::Mapping::new();
    base_entry.insert(Value::String("name".into()), Value::String(distro));
    base_entry.insert(Value::String("channel".into()), Value::String(series));
    base_entry.insert(
        Value::String("architectures".into()),
        Value::Sequence(vec![Value::String(arch.to_string())]),
    );

    manifest.insert(
        "bases".to_string(),
        Value::Sequence(vec![Value::Mapping(base_entry)]),
    );

    let mut analysis = serde_yaml::Mapping::new();
    analysis.insert(
        Value::String("attributes".into()),
        Value::Sequence(vec![]),
    );
    manifest.insert("analysis".to_string(), Value::Mapping(analysis));

    manifest
}

/// Determine the platform label for the charm filename.
fn resolve_platform_label(project: &BTreeMap<String, Value>, arch: &str) -> String {
    if let Some(Value::Mapping(platforms)) = project.get("platforms") {
        for key in platforms.keys() {
            if let Value::String(key_str) = key {
                if key_str == arch {
                    return arch.to_string();
                }
                if key_str.ends_with(&format!(":{arch}"))
                    || key_str.ends_with(&format!("-{arch}"))
                {
                    return key_str.replace(':', "-");
                }
            }
        }
        // Use the first platform key if none matched explicitly.
        if let Some(Value::String(first)) = platforms.keys().next() {
            return first.replace(':', "-");
        }
    }

    let (distro, series) = resolve_base(project);
    format!("{distro}@{series}-{arch}")
}

/// Return the standard charm filename, e.g. `myapp_amd64.charm`.
pub fn charm_filename(project: &BTreeMap<String, Value>, arch: &str) -> String {
    let name = project
        .get("name")
        .and_then(|v| v.as_str())
        .unwrap_or("charm");
    let label = resolve_platform_label(project, arch);
    format!("{name}_{label}.charm")
}

/// Write `actions.yaml` or `config.yaml` into prime_dir.
///
/// Prefers copying the source file from charm_dir if it exists on disk,
/// otherwise generates it from the project dict.
pub fn write_optional_yaml(
    project: &BTreeMap<String, Value>,
    field: &str,
    filename: &str,
    charm_dir: &Path,
    prime_dir: &Path,
) -> Result<(), String> {
    let source = charm_dir.join(filename);
    let dest = prime_dir.join(filename);

    if source.is_file() {
        std::fs::copy(&source, &dest)
            .map_err(|e| format!("Failed to copy {filename}: {e}"))?;
    } else if let Some(val) = project.get(field) {
        if !val.is_null() {
            let content =
                serde_yaml::to_string(val).map_err(|e| format!("Failed to serialize {field}: {e}"))?;
            std::fs::write(&dest, content)
                .map_err(|e| format!("Failed to write {filename}: {e}"))?;
        }
    }

    Ok(())
}
